### PR TITLE
BaSP-TG-08: create and get superAdmins

### PR DESCRIPTION
**Pull request by Maxi-SR**
Get all super admins, get super admin by id and create a super admin. 
Note: I'm working on port 5000 because I'm having problems with 3000, be careful with that.

### CODE REVIEW

File: src/controllers/superAdmins.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    return res.status(200).json({
+      message: 'Super admins found',
+      data: superAdmins,
+      error: false,
+    });
+  } catch (error) {
+    return res.json({
+      message: 'An error ocurred',
+      error,
+    });
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:15]: Please change this error handling with what is proposed in the weekly problem.

File: src/controllers/superAdmins.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      message: `Admin with id: ${idAdmin} found`,
+      data: superAdmin,
+      error: false,
+    });
+  } catch (error) {
+    return res.json({
+      message: 'An error ocurred',
+      error,
+    });
+  }
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:33]: Please change this error handling with what is proposed in the weekly problem.

File: src/controllers/superAdmins.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+const getAdminById = async (req, res) => {
+  const idAdmin = req.params.id;
+  try {
+    const superAdmin = await SuperAdmins.findById(idAdmin);
+    return res.status(200).json({
+      message: `Admin with id: ${idAdmin} found`,
+      data: superAdmin,
+      error: false,
+    });
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:27]: This endpoint is to get a superAdmin by ID. We should update the naming here to reflect that. (Is not to get an admin, but a superAdmin)

Comment (martinraveglia) [reply src/controllers/superAdmins.js:27]: If there is no admin found with given id, what is the result? It should respond with a 404 error

File: src/controllers/superAdmins.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+const getAllSuperAdmins = async (req, res) => {
+  try {
+    const superAdmins = await SuperAdmins.find();
+    return res.status(200).json({
+      message: 'Super admins found',
+      data: superAdmins,
+      error: false,
+    });
+  } catch (error) {
+    return res.json({
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:12]: we're not setting the error response code here

File: src/controllers/superAdmins.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  const idAdmin = req.params.id;
+  try {
+    const superAdmin = await SuperAdmins.findById(idAdmin);
+    return res.status(200).json({
+      message: `Admin with id: ${idAdmin} found`,
+      data: superAdmin,
+      error: false,
+    });
+  } catch (error) {
+    return res.json({
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:29]: we're not setting the error response code here

File: src/controllers/timeSheets.js [hunk 6]
```diff
@@ -1,87 +1,87 @@
-import express from 'express';
-import fs from 'fs';
-import timeSheets from '../models/Timesheets';
+// import express from 'express';
```
Comment (martinraveglia) [inline src/controllers/timeSheets.js:1]: Why is this whole file commented?

File: src/index.js [hunk 7]
```diff
@@ -1,13 +1,13 @@
 import express from 'express';
 import mongoose from 'mongoose';
-import routes from './routes';
+import superAdminRoute from './routes/superAdmins';
 
 const app = express();
-const port = process.env.PORT || 3000;
+const port = process.env.PORT || 5000;
```
Comment (martinraveglia) [inline src/index.js:6]: Please leave this on port 3000, change it In your PC if needed, but don't push those changes

File: src/models/SuperAdmins.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import mongoose from 'mongoose';
+
+const { Schema } = mongoose;
+
+const superAdminSchema = new Schema({
+
```
Comment (martinraveglia) [inline src/models/SuperAdmins.js:6]: please remove this line break

File: src/models/SuperAdmins.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  name: { type: String, required: true },
+  lastName: { type: String, required: true },
+  email: { type: String, required: true },
+  password: { type: String, required: true },
+
```
Comment (martinraveglia) [inline src/models/SuperAdmins.js:11]: please remove this line break

File: src/routes/superAdmins.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+import superAdminsController from '../controllers/superAdmins';
+import superAdminsValidation from '../validations/superAdmins';
+
+const router = express.Router();
+router
+  .get('/', superAdminsController.getAllSuperAdmins)
+  .get('/:id', superAdminsController.getAdminById)
+  .post('/', superAdminsValidation.validateCreation, superAdminsController.createSuperAdmin)
+  .delete('/:id')
+  .put('/:id');
```
Comment (martinraveglia) [inline src/routes/superAdmins.js:11]: Please, update how this is defined, do it as follows to keep the code understandable and more tidy,

```
router.get('/', superAdminsController.getAllSuperAdmins)

router.get('/:id', superAdminsController.getAdminById)

router.post('/', superAdminsValidation.validateCreation, superAdminsController.createSuperAdmin)

router.delete('/:id')

router.put('/:id');
```

File: src/controllers/superAdmins.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+      error,
+    });
+  }
+};
+
+export default {
+  getAllSuperAdmins,
+  getAdminById,
+  createSuperAdmin,
+};
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:63]: Let's export this with no `default`, as 
```
export {
  getAllSuperAdmins,
  getAdminById,
  createSuperAdmin,
};
```

File: src/controllers/superAdmins.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+      message: 'Super admin created succesfully',
+      data: result,
+      error: false,
+    });
+  } catch (error) {
+    return res.status(400).json({
+      message: 'An error ocurred',
+      error,
+    });
+  }
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:56]: Please change this error handling with what is proposed in the weekly problem.

File: src/validations/superAdmins.js [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+      error: true,
+    });
+  }
+
+  return next();
+};
+
+export default {
+  validateCreation,
+};
```
Comment (martinraveglia) [inline src/validations/superAdmins.js:27]: Let's export this with no `default`

File: src/routes/superAdmins.js [hunk 14]
```diff
@@ -0,0 +1,13 @@
+import express from 'express';
+import superAdminsController from '../controllers/superAdmins';
+import superAdminsValidation from '../validations/superAdmins';
```
Comment (martinraveglia) [inline src/routes/superAdmins.js:3]: After changing exports on those files, import them method by method, as 
```
import { validateCreation } from '../validations/superAdmins';
```

File: postman_collection.json [hunk 15]
```diff
@@ -16,7 +16,78 @@
...
+        },
+        "response": []
+      },
+      {
+        "name": "Get super admin by id",
+        "request": {
+          "method": "GET",
+          "header": [],
+          "url": {
+            "raw": "localhost:5000/superAdmins/:id",
```
Comment (martinraveglia) [inline postman_collection.json:72]: Let's keep this on port 3000, change it in your PC if needed, but do not push those port changes

File: postman_collection.json [hunk 16]
```diff
@@ -16,7 +16,78 @@
...
+            "mode": "raw",
+            "raw": "{\n    \"name\":\"admin2\",\n    \"lastName\":\"adminLn2\",\n    \"email\":\"admin2@gmail.com\",\n    \"password\":\"admin1234\"\n}",
+            "options": {
+              "raw": {
+                "language": "json"
+              }
+            }
+          },
+          "url": {
+            "raw": "localhost:5000/superAdmins/",
```
Comment (martinraveglia) [inline postman_collection.json:53]: Let's keep this on port 3000, change it in your PC if needed, but do not push those port changes

File: postman_collection.json [hunk 17]
```diff
@@ -16,7 +16,78 @@
...
 		{
 			"name": "Super admin",
-			"item": []
+			"item": [{
+        "name": "Get all super admins",
+        "request": {
+          "method": "GET",
+          "header": [],
+          "url": {
+            "raw": "localhost:5000/superAdmins/",
```
Comment (martinraveglia) [inline postman_collection.json:25]: Let's keep this on port 3000, change it in your PC if needed, but do not push those port changes

**Review by martinraveglia — CHANGES_REQUESTED** [review]
The name of this PR is not OK, it should be `BaSP-TG-XX: Short description`
The name of the branch is not ok either, it should have been `feature/TG-8/create-get-superAdmins`. Don't change it now, as it's not simple. But take that I mind.

File: src/controllers/superAdmins.js (hunk 3 above)
Comment (Maxi-SR) [reply src/controllers/superAdmins.js:27]: The 404 error has to be in the catch, right?

Comment (Maxi-SR) [reply src/controllers/superAdmins.js:27]: `const getSuperAdminById = async (req, res) => {
  const idSuperAdmin = req.params.id;
  try {
    const superAdmin = await SuperAdmins.findById(idSuperAdmin);
    console.log(superAdmin);
    return res.status(200).json({
      message: `Admin with id: ${idSuperAdmin} found`,
      data: superAdmin,
      error: false,
    });
  } catch (error) {
    return res.status(404).json({
      message: 'An error ocurred',
      error,
    });
  }
};`

Comment (Maxi-SR) [reply src/controllers/superAdmins.js:27]: Like this? 

`const getSuperAdminById = async (req, res) => {
  const idSuperAdmin = req.params.id;
  try {
    const superAdmin = await SuperAdmins.findById(idSuperAdmin);
    console.log(superAdmin);
    return res.status(200).json({
      message: `Super admin with id: ${idSuperAdmin} found`,
      data: superAdmin,
      error: false,
    });
  } catch (error) {
    return res.status(404).json({
      message: `Super admin with id: ${idSuperAdmin} not exist`,
      error,
    });
  }
};`

File: src/controllers/superAdmins.js (hunk 2 above)
Comment (Maxi-SR) [reply src/controllers/superAdmins.js:33]: `} catch (error) {
    return res.status(400).json({
      message: 'An error ocurred',
      error,
    });
  }
};`

File: src/controllers/timeSheets.js (hunk 6 above)
Comment (Maxi-SR) [reply src/controllers/timeSheets.js:1]: Do I have to delete all from this file?

File: src/controllers/superAdmins.js (hunk 1 above)
Comment (Maxi-SR) [reply src/controllers/superAdmins.js:15]: > The name of this PR is not OK, it should be `BaSP-TG-XX: Short description` The name of the branch is not ok either, it should have been `feature/TG-8/create-get-superAdmins`. Don't change it now, as it's not simple. But take that I mind.

I thought it was implied in the feature, sorry for that.

File: src/controllers/superAdmins.js (hunk 12 above)
Comment (Maxi-SR) [reply src/controllers/superAdmins.js:56]: I do not understand what I've to change here? Status code?

File: src/controllers/superAdmins.js (hunk 5 above)
Comment (Maxi-SR) [reply src/controllers/superAdmins.js:29]: Is this the same comment as above?

File: src/controllers/superAdmins.js (hunk 3 above)
Comment (martinraveglia) [reply src/controllers/superAdmins.js:27]: No, as proposed in the weekly exercise.
```
return res.status(404).json({
    message: error, // error.toString()
    data: undefined,
    error: true
});
```

Let's try and define if is better to put `error` or `error.toString()`

File: src/controllers/superAdmins.js (hunk 5 above)
Comment (martinraveglia) [reply src/controllers/superAdmins.js:29]: no, we need to set the statusCode to a 400 or 404 code on every error responses, in this case it should be a 404 code.

File: src/controllers/superAdmins.js (hunk 2 above)
Comment (martinraveglia) [reply src/controllers/superAdmins.js:33]: No, as proposed in the weekly exercise.

```
return res.status(404).json({
    message: error, // error.toString()
    data: undefined,
    error: true
});
```

Let's try and define if is better to put error or error.toString()

File: src/controllers/superAdmins.js (hunk 12 above)
Comment (martinraveglia) [reply src/controllers/superAdmins.js:56]:
```
return res.status(400).json({
    message: error, // error.toString()
    data: undefined,
    error: true
});
```

File: src/controllers/timeSheets.js (hunk 6 above)
Comment (martinraveglia) [reply src/controllers/timeSheets.js:1]: No, we should let it uncommented as it was originally, the teammate who needs to work on the timeSheets related methods will handle this file

**Review by martinraveglia — CHANGES_REQUESTED** [review]

File: src/controllers/superAdmins.js [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    return res.status(200).json({
+      message: 'Super admins found',
+      data: result,
+      error: false,
+    });
+  } catch (error) {
+    let statusCode = 400;
+    if (error.message.includes('Super admin with')) {
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:18]: This will never match the error

File: src/controllers/superAdmins.js [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    });
+
+    const result = await superAdmin.save();
+    return res.status(201).json({
+      message: 'Super admin created succesfully',
+      data: result,
+      error: false,
+    });
+  } catch (error) {
+    return res.status(404).json({
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:66]: this error should not be a 404

File: src/models/SuperAdmins.js [hunk 20]
```diff
@@ -0,0 +1,10 @@
+import mongoose from 'mongoose';
+
+const { Schema } = mongoose;
+const superAdminSchema = new Schema({
```
Comment (martinraveglia) [inline src/models/SuperAdmins.js:4]: A lineBreak is necessary here in between line 3 and 4

File: src/models/SuperAdmins.js [hunk 21]
```diff
@@ -0,0 +1,10 @@
...
+  name: { type: String, required: true },
+  lastName: { type: String, required: true },
+  email: { type: String, required: true },
+  password: { type: String, required: true },
+});
+export default mongoose.model('SuperAdmin', superAdminSchema);
```
Comment (martinraveglia) [inline src/models/SuperAdmins.js:10]: A lineBreak is necessary here in between line 9 and 10

**Review by martinraveglia — CHANGES_REQUESTED** [review]

File: src/controllers/superAdmins.js [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+    if (superAdmin === null) {
+      throw new Error('Super admin not found');
+    }
+    return res.status(200).json({
+      message: `Super admin with id: ${idSuperAdmin} found`,
+      data: superAdmin,
+      error: false,
+    });
+  } catch (error) {
+    return res.status(404).json({
```
Comment (martinraveglia) [inline src/controllers/superAdmins.js:42]: statusCode should be 404 only if no superAdmin was found with the given Id, it should be 400 otherwise

**Review by martinraveglia — CHANGES_REQUESTED** [review]

File: src/index.js [hunk 23]
```diff
@@ -1,13 +1,13 @@
...
-import routes from './routes';
+import superAdminRoute from './routes/superAdmins';
 
 const app = express();
 const port = process.env.PORT || 3000;
 const MONGO_URL = 'mongodb+srv://RadiumA:RadiumA@trackgenix.r6u6do6.mongodb.net/?retryWrites=true&w=majority';
 
 app.use(express.json());
-app.use(routes);
+app.use('/superAdmins', superAdminRoute);
```
Comment (martinraveglia) [inline src/index.js:10]: This belongs to the routes index file!!

**Review by martinraveglia — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 23 above)
Comment (Maxi-SR) [reply src/index.js:10]: I think I got it right, I'm not sure but it works

File: src/index.js [hunk 24]
```diff
@@ -1,13 +1,13 @@
 import express from 'express';
 import mongoose from 'mongoose';
-import routes from './routes';
+import router from './routes/index';
```
Comment (martinraveglia) [inline src/index.js:3]: this should be named routes (as in the original file)

File: src/validations/superAdmins.js [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+
+const validateCreation = (req, res, next) => {
+  const pattern = /^[^@]+@[^@]+\.[a-zA-Z]{2,}$/;
+  const superAdminValidation = Joi.object({
+    name: Joi.string().min(3).max(50).required(),
+    lastName: Joi.string().min(3).max(50).required(),
+    email: Joi.string().pattern(new RegExp(pattern)).required(),
+    password: Joi.string().alphanum().min(8).required(),
+  });
+  const validation = superAdminValidation.validateAsync(req.body, { abortEarly: false });
```
Comment (martinraveglia) [inline src/validations/superAdmins.js:11]: why are you using validateAsync? This calls for an asynchronous method that you're not handling and therefore the app crashes when validation is not successful. Let's use `.validate` instead as it is synchronous.

File: src/validations/superAdmins.js [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+  if (validation.error) {
+    return res.status(400).json({
+      message: `There was an error: ${validation.error.details[0].message}`,
+      data: undefined,
+      error: true,
+    });
```
Comment (martinraveglia) [inline src/validations/superAdmins.js:17]: if we have more than one validation error (in more than one field), all errors should be sent

**Review by martinraveglia — CHANGES_REQUESTED** [review]

File: src/index.js [hunk 27]
```diff
@@ -1,6 +1,6 @@
 import express from 'express';
 import mongoose from 'mongoose';
-import routes from './routes';
+import routes from './routes/index';
```
Comment (martinraveglia) [inline src/index.js:3]:
```suggestion
import routes from './routes';
```

**Review by martinraveglia — CHANGES_REQUESTED** [review]

File: src/routes/index.js [hunk 28]
```diff
@@ -1,6 +1,8 @@
 import express from 'express';
 import superAdmins from './superAdmins';
 // import admins from './admins';
```
Comment (martinraveglia) [inline src/routes/index.js:3]: please delete this line

File: src/routes/index.js [hunk 29]
```diff
@@ -12,6 +14,8 @@ const router = express.Router();
 
 router.use('/super-admins', superAdmins);
 // router.use('/admins', admins);
```
Comment (martinraveglia) [inline src/routes/index.js:16]: please delete this line

**Review by martinraveglia — CHANGES_REQUESTED** [review]

**Review by martinraveglia — APPROVED** [review]

**Pull request merged**